syntax = "proto3";

package vectron.v1;

import "google/api/annotations.proto";
import "google/api/field_behavior.proto";

option go_package = "github.com/pavandhadge/vectron/clientlibs/js/proto/apigateway";

// ===============================
// Vectron Public API
// ===============================

service VectronService {
  rpc CreateCollection(CreateCollectionRequest) returns (CreateCollectionResponse) {
    option (google.api.http) = {
      post: "/v1/collections"
      body: "*"
    };
  }

  rpc Upsert(UpsertRequest) returns (UpsertResponse) {
    option (google.api.http) = {
      post: "/v1/collections/{collection}/points"
      body: "*"
    };
  }

  rpc Search(SearchRequest) returns (SearchResponse) {
    option (google.api.http) = {
      post: "/v1/collections/{collection}/points/search"
      body: "*"
    };
  }

  rpc Get(GetRequest) returns (GetResponse) {
    option (google.api.http) = {
      get: "/v1/collections/{collection}/points/{id}"
    };
  }

  rpc Delete(DeleteRequest) returns (DeleteResponse) {
    option (google.api.http) = {
      delete: "/v1/collections/{collection}/points/{id}"
    };
  }

  rpc ListCollections(ListCollectionsRequest) returns (ListCollectionsResponse) {
    option (google.api.http) = {
      get: "/v1/collections"
    };
  }
}

// ===============================
// Messages
// ===============================

message CreateCollectionRequest {
  string name = 1 [(google.api.field_behavior) = REQUIRED];
  int32 dimension = 2 [(google.api.field_behavior) = REQUIRED];
  string distance = 3;
}

message CreateCollectionResponse {
  bool success = 1;
}

message UpsertRequest {
  string collection = 1 [(google.api.field_behavior) = REQUIRED];
  repeated Point points = 2;
}

message UpsertResponse {
  int32 upserted = 1;
}

message SearchRequest {
  string collection = 1 [(google.api.field_behavior) = REQUIRED];
  repeated float vector = 2 [(google.api.field_behavior) = REQUIRED];
  uint32 top_k = 3;
}

message SearchResponse {
  repeated SearchResult results = 1;
}

message SearchResult {
  string id = 1;
  float score = 2;
  map<string, string> payload = 3;
}

message GetRequest {
  string collection = 1;
  string id = 2;
}

message GetResponse {
  Point point = 1;
}

message DeleteRequest {
  string collection = 1;
  string id = 2;
}

message DeleteResponse {}

message ListCollectionsRequest {}

message ListCollectionsResponse {
  repeated string collections = 1;
}

message Point {
  string id = 1;
  repeated float vector = 2;
  map<string, string> payload = 3;
}
